src/Stratis.Bitcoin/AsyncWork/AsyncProvider.cs

```diff
@@ -0,0 +1,318 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+using TracerAttributes;
+
+namespace Stratis.Bitcoin.AsyncWork
+{
+    /// <summary>
+    /// Provides functionality for creating and tracking asynchronous operations that happens in background.
+    /// </summary>
+    public partial class AsyncProvider : IAsyncProvider
+    {
+        private const int DefaultLoopRepeatInterval = 1000;
+
+        private object lockAsyncDelegates;
+
+        /// <summary>
+        /// Holds a list of currently running async delegates or delegates that stopped because of unhandled exceptions.
+        /// Protected by <see cref="lockAsyncDelegates"/> lock
+        /// </summary>
+        private Dictionary<IAsyncDelegate, AsyncTaskInfo> asyncDelegates;
+
+        private ILoggerFactory loggerFactory;
+        private ILogger logger;
+        private ISignals signals;
+        private readonly INodeLifetime nodeLifetime;
+
+        private (string Name, int Width)[] benchmarkColumnsDefinition = new[]
+        {
+            (Name: "Name", Width: 80),
+            (Name: "Type", Width: 15),
+            (Name: "Health", Width: 15)
+        };
+
+        public AsyncProvider(ILoggerFactory loggerFactory, ISignals signals, INodeLifetime nodeLifetime)
+        {
+            this.lockAsyncDelegates = new object();
+
+            this.asyncDelegates = new Dictionary<IAsyncDelegate, AsyncTaskInfo>();
+
+            this.loggerFactory = Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            this.logger = this.loggerFactory.CreateLogger(nameof(AsyncProvider));
+
+            this.signals = Guard.NotNull(signals, nameof(signals));
+            this.nodeLifetime = Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+        }
+
+        /// <inheritdoc />
+        public IAsyncDelegateDequeuer<T> CreateAndRunAsyncDelegateDequeuer<T>(string friendlyName, Func<T, CancellationToken, Task> @delegate)
```

Comment:
Should we not limit to have only one instance by name?

Reply:
Ah we cant because of the behaviour right?

Reply:
yes there are actually scenarios where the name is duplicated, I wouldn't enforce it to be unique (even if having an unique name help to debug easier)

I'd say who uses this class should try to give a meaningful name
I think I could add a Rename method to AsyncProvider, to update the name when is possible, to have a better name after the constructor/initialization, but maybe it's not that important